waku/README.md

```diff
@@ -0,0 +1,301 @@
+
+It is only relevant if you want to decrypt the incoming message, but if you only want to send a message, any other format would be perfectly valid and must be forwarded to the peers.
+
+Data field contains encrypted message of the Envelope. In case of symmetric encryption, it also contains appended Salt (a.k.a. AES Nonce, 12 bytes). Plaintext (unencrypted) payload consists of the following concatenated fields: flags, auxiliary field, payload, padding and signature (in this sequence).
+
+	flags: 1 byte; first two bits contain the size of auxiliary field, third bit indicates whether the signature is present.
+	
+	auxiliary field: up to 4 bytes; contains the size of payload.
+
+	payload: byte array of arbitrary size (may be zero).
+
+	padding: byte array of arbitrary size (may be zero).
+
+	signature: 65 bytes, if present.
+
+	salt: 12 bytes, if present (in case of symmetric encryption).
+
+Those unable to decrypt the message data are also unable to access the signature. The signature, if provided, is the ECDSA signature of the Keccak-256 hash of the unencrypted data using the secret key of the originator identity. The signature is serialised as the concatenation of the `R`, `S` and `V` parameters of the SECP-256k1 ECDSA signature, in that order. `R` and `S` are both big-endian encoded, fixed-width 256-bit unsigned. `V` is an 8-bit big-endian encoded, non-normalised and should be either 27 or 28.
+
+The padding field was introduced in order to align the message size, since message size alone might reveal important metainformation. Padding can be arbitrary size. However, it is recommended that the size of Data Field (excuding the Salt) before encryption (i.e. plain text) should be factor of 256 bytes.
+
+### Payload Encryption
+
+Asymmetric encryption uses the standard Elliptic Curve Integrated Encryption Scheme with SECP-256k1 public key.
+
+Symmetric encryption uses AES GCM algorithm with random 96-bit nonce.
+
+## Additional capabilities
+
+Waku supports multiple capabilities. These include light node, rate limting, mailserver (client and server) and bridging of traffic. Here we list these capabilities, how they are identified, what properties they have and what invariants they must maintain.
+
+### Light node
+
+The rationale for light nodes is to allow for interaction with waku on resource restricted devices as bandwidth can often be an issue.
+
+Light nodes MUST NOT forward any incoming messages, they MUST only send their own messages. Light nodes SHOULD not connect to other light nodes, this would result in messages being dropped between the two.
+
+Light nodes are identified using the handshake.
+
+<!-- TODO: Add details on handshake -- >
+
+**Security consideration**:
+Running a node as a light node mode impacts privacy due to the fact that it becomes identifiable what nodes care about if they aren't relaying traffic.
```

Comment:
Is this true? I don't think it is completely correct. If a light node only listens and never sends any message, and it could even do this with a full bloom, then you will not know what it cares about.

I think the main privacy concern with light nodes is that the directly connected peers will know that a message originates from them (as it are the only ones it sends). And yes, based on that they can make some assumptions on which messages (topics) they are interested in also.

Side note: I only recently came to learn that the geth/status light node implementation allows for sending messages. In nim-eth this is not allowed, unless perhaps when you send p2p messages to a trusted peer. I believe this was the original intention..?

Reply:
@kdeme the problem is most light node clients are usually bandwidth restricted so listening to everything would make no sense as you would essentially be DoSing yourself by doing so.

Reply:
> I think the main privacy concern with light nodes is that the directly connected peers will know that a message originates from them (as it are the only ones it sends). And yes, based on that they can make some assumptions on which messages (topics) they are interested in also.

True, fixed (commented out for now, separate issue)

> Side note: I only recently came to learn that the geth/status light node implementation allows for sending messages. In nim-eth this is not allowed, unless perhaps when you send p2p messages to a trusted peer. I believe this was the original intention..?

How else would a mobile user send messages? :D Do you mean relay?

Reply:
Well, I honestly never looked in great detail into the light mode of geth, only skimmed it. As there was/is no spec at all and I assumed it to be a solution still in flux (and thought it was not really used).

That said, in nim-eth we do not allow for sending like normal messages due to privacy considerations, but it is still possible to send. Because I noticed there is this `trusted peer` flag (and `setPeerTrusted` function), and there is the `p2pMessage` packet id, that combination is what is being used for light nodes. So it is still possible to send messages, just not in the regular way (this is the way I assumed/interpreted how geth did it also). 
See e.g. here: https://github.com/status-im/nim-eth/blob/master/eth/p2p/rlpx_protocols/whisper_protocol.nim#L345 (it requires a `targetPeer` that is `trusted`). I think a similar system is used for mailservers (but with the `p2pRequest` message).

Reply:
Interesting point - this points to the merits of a good spec! The trusted peer and mark trusted is only used for mailservers afaik. Light nodes should be able to send normally, just not relay traffic. At least that's how I've always understood it. @adambabik can probably confirm